File: src/first_cpp_pkg/src/service_server.cpp
#include <chrono>
#include <memory>
#include <vector>
#include <cmath>
#include <unistd.h>
#include "rclcpp/rclcpp.hpp"
#include "first_package_msgs/srv/multi_spawn.hpp"
#include "turtlesim/srv/spawn.hpp"
#include "turtlesim/srv/teleport_absolute.hpp"


using std::placeholders::_1;
using std::placeholders::_2;

class MultiSpawning : public rclcpp::Node{
public:
    MultiSpawning() : Node("multi_spawn")
    {
        server = this->create_service<first_package_msgs::srv::MultiSpawn>(
            "multi_spawn", std::bind(&MultiSpawning::service_callback, this, _1, _2));

        client_spawn = this->create_client<turtlesim::srv::Spawn>("spawn");
        client_teleport = this->create_client<turtlesim::srv::TeleportAbsolute>("turtle1/teleport_absolute");
        

    }
private:
    std::vector<std::vector<float>> calc_position(int num, float r){
        std::vector<std::vector<float>> result(3, std::vector<float>(num));
        float gap_theta = 2 * M_PI / num;
        for (int i = 0; i < num; ++i) {
            float theta = i * gap_theta;
            result[0][i] = r * std::cos(theta); // x 좌표
            result[1][i] = r * std::sin(theta); // y 좌표
            result[2][i] = theta; // 각도값
        }
        return result;
    }
    void service_callback(const std::shared_ptr<first_package_msgs::srv::MultiSpawn::Request> req,
    const std::shared_ptr<first_package_msgs::srv::MultiSpawn::Response> res){
        int num = req->num;
        float rad = 3; // 반지름
        auto client_spawn_req = std::make_shared<turtlesim::srv::Spawn::Request>();
        std::vector<std::vector<float>> positions = calc_position(num, rad);


        for (int i = 0; i < num; ++i){
            client_spawn_req->x = positions[0][i] + 5.54;
            client_spawn_req->y = positions[1][i] + 5.54;
            client_spawn_req->theta = positions[2][i];

            auto result_future = client_spawn->async_send_request(client_spawn_req);
            
            usleep(200000);
        }
        

    }


    rclcpp::Service<first_package_msgs::srv::MultiSpawn>::SharedPtr server;
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr client_spawn;
    rclcpp::Client<turtlesim::srv::TeleportAbsolute>::SharedPtr client_teleport;

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MultiSpawning>());
    rclcpp::shutdown();
    return 0;
}

File: src/first_package/first_package/service_server.py
import rclpy as rp
from rclpy.node import Node
import os
import time
import numpy as np
from turtlesim.srv import TeleportAbsolute
from turtlesim.srv import Spawn

from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
from first_package_msgs.srv import MultiSpawn



class MultiSpawning(Node) : 
    def __init__(self) :
        super().__init__("multi_spawn")
        self.server = self.create_service(MultiSpawn, "multi_spawn", self.callback_service)
        self.teleport = self.create_client(TeleportAbsolute, "/turtle1/teleport_absolute")
        self.spawn = self.create_client(Spawn, "/spawn")
        self.req_spawn = Spawn.Request()
        self.center_x = 5.54
        self.center_y = 5.54
        
        

    def callback_service(self, req, res) : 
        x, y, theta = self.calc_pos(req.num, 3)

        for n in range(len(theta)):
            self.req_spawn.x = x[n] + self.center_x
            self.req_spawn.y = y[n] + self.center_y
            self.req_spawn.theta = theta[n]
            self.spawn.call_async(self.req_spawn)
            
            time.sleep(0.1)
        res.x = x
        res.y = y
        res.theta = theta
        
        

        return res
    
    def calc_pos(self, n, r) :
        gap_theta = 2*np.pi / n
        theta = [gap_theta*n for n in range(n)]
        x = [r*np.cos(th) for th in theta]
        y = [r*np.sin(th) for th in theta]
        
        return x, y, theta

def main(args=None) : 
    rp.init(args=args)

    service = MultiSpawning()
    rp.spin(service)

    service.destroy_node()
    rp.shutdown()


if __name__ == "__main__" :
    main()